### ISSUE TITLE
`org-cite-make-insert-processor`: Wrong argument type(s)

### ISSUE THREAD

**Issue by wuqui** [bdarcus/citar#352]
**Describe the bug**

I wanted to try the new configuration for `oc-bibtex-actions` I found on the new pull request for Doom's `biblio` module. I added the new code for `config.el` and `packages.el` to my config. However, the following block causes problems:

```
(use-package! oc-bibtex-actions
  :when (featurep! :completion vertico)
  :after oc
  :defer t
  :init
  ;; While this processor is registered within 'oc-bibtex-actions', we do it here
  ;; to avoid having to load it upfront.
  (org-cite-register-processor 'oc-bibtex-actions
    :insert (org-cite-make-insert-processor
             #'oc-bibtex-actions-insert
             #'oc-bibtex-actions-select-style)
    :follow #'oc-bibtex-actions-follow
    :activate #'oc-bibtex-actions-activate)
  (setq org-cite-insert-processor 'oc-bibtex-actions
        org-cite-follow-processor 'oc-bibtex-actions
        org-cite-activate-processor 'oc-bibtex-actions
        ;; The activate processor relies on shift-select, so we set to t.
        org-support-shift-select t)
  :config
  (when (featurep! :lang org +roam2)
    ;; Include property drawer metadata for 'org-roam' v2.
    (setq bibtex-actions-file-note-org-include '(org-id org-roam-ref))))
```

I can't use it since I get the following error when starting Emacs:

```
Debugger entered--Lisp error: (error "Wrong argument type(s)")
  signal(error ("Wrong argument type(s)"))
  error("Wrong argument type(s)")
  org-cite-make-insert-processor(oc-bibtex-actions-insert oc-bibtex-actions-select-style)
  (org-cite-register-processor 'oc-bibtex-actions :insert (org-cite-make-insert-processor #'oc-bibtex-actions-insert #'oc-bibt$
  (progn (use-package-statistics-gather :init 'oc-bibtex-actions nil) (org-cite-register-processor 'oc-bibtex-actions :insert $
  (lambda nil (progn (use-package-statistics-gather :init 'oc-bibtex-actions nil) (org-cite-register-processor 'oc-bibtex-acti$
  eval-after-load(oc (lambda nil (progn (use-package-statistics-gather :init 'oc-bibtex-actions nil) (org-cite-register-proces$
  (progn (eval-after-load 'oc #'(lambda nil (progn (use-package-statistics-gather :init 'oc-bibtex-actions nil) (org-cite-regi$
  (if t (progn (eval-after-load 'oc #'(lambda nil (progn (use-package-statistics-gather :init 'oc-bibtex-actions nil) (org-cit$
  eval-buffer(#<buffer  *load*-235726> nil "/Users/quirin/.doom.d/config.el" nil t)  ; Reading at buffer position 16203
  load-with-code-conversion("/Users/quirin/.doom.d/config.el" "/Users/quirin/.doom.d/config.el" t t)
  load("/Users/quirin/.doom.d/config" t nomessage)
  (let (file-name-handler-alist) (load (expand-file-name "config" doom-private-dir) t 'nomessage))
  (condition-case e (let (file-name-handler-alist) (load (expand-file-name "config" doom-private-dir) t 'nomessage)) ((debug d$
  (if no-config-p nil (maphash (doom-module-loader doom-module-config-file) doom-modules) (doom-run-hooks 'doom-init-modules-h$
  (progn (if doom-debug-p (progn (let ((inhibit-message (active-minibuffer-window))) (message #("DOOM Initializing user config$
  (if init-p (progn (if doom-debug-p (progn (let ((inhibit-message (active-minibuffer-window))) (message #("DOOM Initializing $
  (let* ((init-p (and t (condition-case e (let (file-name-handler-alist) (load (expand-file-name doom-module-init-file doom-pr$
  (progn (setq doom-init-modules-p t) (if no-config-p nil (if doom-debug-p (progn (let ((inhibit-message (active-minibuffer-wi$
  (if (or force-p (not doom-init-modules-p)) (progn (setq doom-init-modules-p t) (if no-config-p nil (if doom-debug-p (progn ($
  doom-initialize-modules()
  eval-buffer(#<buffer  *load*> nil "/Users/quirin/.emacs.d/init.el" nil t)  ; Reading at buffer position 2804
  load-with-code-conversion("/Users/quirin/.emacs.d/init.el" "/Users/quirin/.emacs.d/init.el" t t)
  load("/Users/quirin/.emacs.d/init" noerror nomessage)
  startup--load-user-init-file(#f(compiled-function () #<bytecode 0x4aa5c905>) #f(compiled-function () #<bytecode 0x4aa5c90f>)$
  command-line()
```

Any ideas what I need to do to fix this?

By the way, thanks for your effort for `org-cite` and the ecosystem. Having a great, native solution for references in `org-mode` will be a gamechanger 🥳

**Emacs version:**

My `doom/info`: https://pastebin.com/PdtE09pb

**Comment by bdarcus** [comment]
It looks like an error that drove me crazy a few months back that is some subtly of the load order.

@tshu-w - can you look at this?

**Comment by wuqui** [comment]
By the way, here's a link to the relevant section in my config in case that could be helpful: https://github.com/wuqui/emacs/blob/main/config.org#org-cite-et-al

**Comment by bdarcus** [comment]
> By the way, here's a link to the relevant section in my config in case that could be helpful: https://github.com/wuqui/emacs/blob/main/config.org#org-cite-et-al

If you comment out the make-processor llnes, as you have there, does it work correctly?

**Comment by wuqui** [comment]
> If you comment out the make-processor llnes, as you have there, does it work correctly?

Yes, I can load the package and I have basic functionality, but many things don't work. I don't know but I thought it might be because I can't uncomment these lines.

What works, 

1. is that I can indeed cite using the processors. 
1. Plus, I can use, e.g., `bibtex-actions-insert-citation` and get a nice vertico selection menu, and I can then on top use the basic `embark` actions such as jumping to the bib file.

However, the following things don't work:

1. When I press return on a candidate, only `ebib:Schmid2016` gets inserted in the buffer.
1. When I call `org-cite-insert`, it breaks with this error: `user-error: Unknown processor oc-bibtex-actions`
1. I don't get the bibtex-relevant embark options when I call `embark-act` with the cursor over the (correctly highlighted) cite key.
2. I can't jump to my `org-roam` notes using `embark-act`.

**Comment by bdarcus** [comment]
Right, because it defers loading of the needed functions.

Sigh ...

**Comment by tshu-w** [comment]
> can you look at this? I suspect your solution may not work consistently.

@bdarcus Sorry, everything work fine on my side, and I may not be particularly free to try it on Doom (I'll try it if I have time, but I can't guarantee it). I remember it works on your side, then there may be a conflict somewhere.

@wuqui I took a brief look at your configand found that you are using the :after keyword aggressively and incorrectly.

Firstly, 
```emacs-lisp
(after! oc
  (setq org-cite-export-processors '((latex biblatex)
                                     (t csl)
                                     (html csl))))
```
equal to 
```emacs-lisp
(use-package! oc
  :defer t
  :config
  (setq org-cite-export-processors
        '((latex biblatex)
          (t csl)
          (html csl))))
```
you don't need to config twice.

Secondly,
```emacs-lisp
(use-package! oc
  :after org bibtex-completion bibtex-actions
  :config)
```
is equal to 
```emacs-lisp
(with-eval-after-load 'org
  (with-eval-after-load 'bibtex-completion
    (with-eval-after-load 'bibtex-actions
      ;; :init section
      (require 'oc)
      ;; :config section
      )))
```
make oc config work only after bibtex-completion loaded and bibtex-actions doesn't make sence. The correct order is to require bibtex-completion bibtex-actions (If not) when using oc and shouldn't be done by user.

Also your `bibtex-actions` is after `embark`, there will definitely be cases where the oc is needed before the embark load.

*Suggestion*
1. don't use `:after` so aggressive if you don't understand
2. put cursor before `(use-package !` and run `pp-macroexpand-expression` to check if you are defer right.

can you try:
```emacs-lisp
(use-package bibtex-completion
  :defer t
  :config
  (setq bibtex-completion-additional-search-fields '(keywords)
        bibtex-completion-pdf-field "file"));; This tell bibtex-completion to look at the File field of the bibtex to figure out which pdf to open

(use-package ivy-bibtex
  :when (featurep! :completion ivy)
  :defer t
  :config
  (add-to-list 'ivy-re-builders-alist '(ivy-bibtex . ivy--regex-plus)))

(use-package bibtex-actions
  :when (featurep! :completion vertico)
  :defer t
  :config
  (add-to-list 'embark-keymap-alist '(bib-reference . bibtex-actions-map))
  (when (featurep! +roam2)
    setq bibtex-actions-file-note-org-include '(org-id org-roam-ref)))

(use-package citeproc :defer t)

;;; Org-Cite configuration

(use-package oc
  :after org
  :config
  (require 'ox)
  (map! :map org-mode-map
        :localleader
        :desc "Insert citation" "@" #'org-cite-insert)
  (setq org-cite-global-bibliography
        (let ((paths (or bibtex-actions-bibliography
                         bibtex-completion-bibliography)))
          ;; Always return bibliography paths as list for org-cite.
          (if (stringp paths) (list paths) paths)))
  ;; setup export processor; default csl/citeproc-el, with biblatex for latex
  (setq org-cite-export-processors
        '((latex biblatex)
          (t csl)
          (html csl))))

  ;;; Org-cite processors
(use-package oc-biblatex :after oc)

(use-package oc-csl :after oc)

(use-package oc-natbib :after oc)

;;;; Third-party

(use-package oc-bibtex-actions
  :when (featurep! :completion vertico)
  :after oc
  :defer t
  :init
  ;; While this processor is registered within 'oc-bibtex-actions', we do it here
  ;; to avoid having to load it upfront.

  ;; the following lines produce the current error
  (org-cite-register-processor 'oc-bibtex-actions
    :insert (org-cite-make-insert-processor
             #'oc-bibtex-actions-insert
             #'oc-bibtex-actions-select-style)
    :follow #'oc-bibtex-actions-follow
    :activate #'oc-bibtex-actions-activate)

  (setq org-cite-insert-processor 'oc-bibtex-actions
        org-cite-follow-processor 'oc-bibtex-actions
        org-cite-activate-processor 'oc-bibtex-actions
        ;; The activate processor relies on shift-select, so we set to t.
        org-support-shift-select t)
  :config
  (when (featurep! :lang org +roam2)
    ;; Include property drawer metadata for 'org-roam' v2.
    (setq bibtex-actions-file-note-org-include '(org-id org-roam-ref))))
```

**Comment by bdarcus** [comment]
Looks like #353 should fix this.

**Comment by wuqui** [comment]
Thanks for looking into this!

> @wuqui I took a brief look at your configand found that you are using the :after keyword aggressively and incorrectly.

Oh yeah, sorry for that. I have to admit I don't know enough about `use-package` and have to read up on that. Part of the reason this is so messy is that I was desperately trying to change things because the normal approach didn't work.

I have deleted my config now and replaced it with your suggested version (https://pastebin.com/RyV61SMk). Unfortunately, I still get the same error though. Here's the output when starting Emacs with `--debug-init`:

```
Debugger entered--Lisp error: (error "Wrong argument type(s)")
  signal(error ("Wrong argument type(s)"))
  error("Wrong argument type(s)")
  org-cite-make-insert-processor(oc-bibtex-actions-insert oc-bibtex-actions-select-style)
  (org-cite-register-processor 'oc-bibtex-actions :insert (org-cite-make-insert-processor #'oc-bibtex-actions-insert #'oc-bibt$
  (progn (use-package-statistics-gather :init 'oc-bibtex-actions nil) (org-cite-register-processor 'oc-bibtex-actions :insert $
  (lambda nil (progn (use-package-statistics-gather :init 'oc-bibtex-actions nil) (org-cite-register-processor 'oc-bibtex-acti$
  eval-after-load(oc (lambda nil (progn (use-package-statistics-gather :init 'oc-bibtex-actions nil) (org-cite-register-proces$
  (progn (eval-after-load 'oc #'(lambda nil (progn (use-package-statistics-gather :init 'oc-bibtex-actions nil) (org-cite-regi$
  (if t (progn (eval-after-load 'oc #'(lambda nil (progn (use-package-statistics-gather :init 'oc-bibtex-actions nil) (org-cit$
  eval-buffer(#<buffer  *load*-768210> nil "/Users/quirin/.doom.d/config.el" nil t)  ; Reading at buffer position 18958
  load-with-code-conversion("/Users/quirin/.doom.d/config.el" "/Users/quirin/.doom.d/config.el" t t)
  load("/Users/quirin/.doom.d/config" t nomessage)
  (let (file-name-handler-alist) (load (expand-file-name "config" doom-private-dir) t 'nomessage))
  (condition-case e (let (file-name-handler-alist) (load (expand-file-name "config" doom-private-dir) t 'nomessage)) ((debug d$
  (if no-config-p nil (maphash (doom-module-loader doom-module-config-file) doom-modules) (doom-run-hooks 'doom-init-modules-h$
  (progn (if doom-debug-p (progn (let ((inhibit-message (active-minibuffer-window))) (message #("DOOM Initializing user config$
  (if init-p (progn (if doom-debug-p (progn (let ((inhibit-message (active-minibuffer-window))) (message #("DOOM Initializing $
  (let* ((init-p (and t (condition-case e (let (file-name-handler-alist) (load (expand-file-name doom-module-init-file doom-pr$
  (progn (setq doom-init-modules-p t) (if no-config-p nil (if doom-debug-p (progn (let ((inhibit-message (active-minibuffer-wi$
  (if (or force-p (not doom-init-modules-p)) (progn (setq doom-init-modules-p t) (if no-config-p nil (if doom-debug-p (progn ($
  doom-initialize-modules()
  eval-buffer(#<buffer  *load*> nil "/Users/quirin/.emacs.d/init.el" nil t)  ; Reading at buffer position 2804
  load-with-code-conversion("/Users/quirin/.emacs.d/init.el" "/Users/quirin/.emacs.d/init.el" t t)
  load("/Users/quirin/.emacs.d/init" noerror nomessage)
  startup--load-user-init-file(#f(compiled-function () #<bytecode 0x4f64ba05>) #f(compiled-function () #<bytecode 0x4f64ba0f>)$
  command-line()
```

**Comment by bdarcus** [comment]
You might wait just a bit for me to merge #355, and then update.

**Comment by wuqui** [comment]
> You might wait just a bit for me to merge #355, and then update.

Okay 👍

**Comment by bdarcus** [comment]
I just merged #355.

I'll update the doom PR, but you should be able to adapt the update README config example?

**Comment by tshu-w** [comment]
@wuqui issue is becasue emacs doesn't know oc-bibtex-actions-insert or oc-bibtex-actions-select-style. Which commit are you using, you can change to citar and try the latest commit.

**Comment by bdarcus** [comment]
Yes, why I was telling him to update.

**Comment by bdarcus** [comment]
@wuqui - this working?

**Comment by wuqui** [comment]
I'm currently swamped with work so couldn't try yet, unfortunately. It's high up my org list though 😉 I'll let you know!

**Comment by bdarcus** [comment]
When you do, see updated READMEz which reflects recent changes.

I'm going to close this for now. If your problem persists, feel free to reopen.

**Comment by wuqui** [comment]
I hope I can try again tomorrow. Thanks a lot for your support, I'll let you know!

**Comment by wuqui** [comment]
Sorry for taking so long. I tried again and everything works beautifully now using what you provide in the Readme. And my config is much leaner now, too.

The only thing that isn't working yet is the integration with `org-roam`. I couldn't figure that out yet. But I think you're also working on integrating all of this in the `biblio` module for Doom Emacs, so maybe I'll just wait till that's ready for off-the-shelf use.

Thanks again for your help!

**Comment by bdarcus** [comment]
> The only thing that isn't working yet is the integration with `org-roam`. I couldn't figure that out yet. But I think you're also working on integrating all of this in the `biblio` module for Doom Emacs, so maybe I'll just wait till that's ready for off-the-shelf use.

The Doom 5290 PR is basically what I'm using; should work fine with `org +roam2`.

Really just waiting on Henrik to find time to review.
